examples/component/http-keyvalue-crud/main.go

```diff
@@ -0,0 +1,185 @@
+
+	// The cm module provides types and functions for interacting with the WebAssembly Component Model.
+	"go.bytecodealliance.org/cm"
+
+	// The wasmCloud wasihttp module enables us to write more idiomatic Go when using wasi:http.
+	"go.wasmcloud.dev/component/net/wasihttp"
+)
+
+// Types for JSON validation.
+type CheckRequest struct {
+	Value string `json:"value"`
+}
+
+type CheckResponse struct {
+	Valid   bool   `json:"valid"`
+	Length  int    `json:"length,omitempty"`
+	Message string `json:"message,omitempty"`
+}
+
+func init() {
+	// Establishes the routes and methods for our keyvalue operations.
+	router := httprouter.New()
+	router.HandlerFunc(http.MethodGet, "/", indexHandler)
+	router.HandlerFunc(http.MethodPost, "/post", postHandler)
+	router.HandlerFunc(http.MethodGet, "/get", getHandler)
+	router.HandlerFunc(http.MethodDelete, "/delete", deleteHandler)
+	wasihttp.Handle(router)
+}
+
+func indexHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Fprintf(w, "Provide a key as query string to /post, /get, or /delete with the corresponding method\n")
+}
+
+func postHandler(w http.ResponseWriter, r *http.Request) {
+	// Checks for a valid key as query string.
+	// For all routes and methods, the user will provide the key via a query string, for example:
+	// curl 'localhost:8000/get?key'
+	if len(r.URL.RawQuery) > 0 && regexp.MustCompile(`^[a-zA-Z0-9]*$`).MatchString(r.URL.RawQuery) {
+
+		// Assigns the raw query string to the "key" variable.
+		key := r.URL.RawQuery
+
+		// Checks the request for a valid JSON body and assigns it to the value variable.
+		// The user will set the value via JSON payload:
+		// curl -X POST 'localhost:8000/post?key' -d '{"foo": "bar", "woo": "hoo"}'
+		var req CheckRequest
+		defer r.Body.Close()
+		value, err := io.ReadAll(r.Body)
+		if err != nil {
+			errResponseJSON(w, http.StatusBadRequest, err.Error())
+			return
+		}
+		if err := json.Unmarshal(value, &req); err != nil {
+			errResponseJSON(w, http.StatusBadRequest, fmt.Sprintf("error with json input: %s", err.Error()))
+			return
+		}
+
+		// Opens the keyvalue bucket.
+		kvStore := store.Open("default")
+		if err := kvStore.Err(); err != nil {
+			w.Write([]byte("Error: " + err.String()))
+			return
+		}
+
+		// Converts the value to a byte array.
+		valueBytes := []byte(value)
+
+		// Converts the byte array to the Component Model's cm.List type.
+		valueList := cm.ToList(valueBytes)
+
+		// Sets the value for the key in the current bucket and handles any errors.
+		kvSet := store.Bucket.Set(*kvStore.OK(), key, valueList)
+		if err := kvSet.Err(); err != nil {
+			w.Write([]byte("Error: " + err.String()))
+			return
+		} else {
+			// Confirms that the key has been set.
+			fmt.Fprintf(w, "Set %s to %s\n", key, value)
+		}
+
+	} else {
+
+		// Prompts the user if no valid key is entered.
+		fmt.Fprintf(w, "POST a key as an alphanumeric query with a JSON body\n")
+	}
```

Comment:
As a general rule of thumb, in Go it's better to return early when a pre-condition isn't met, rather than doing nested-if statements whenever you can help it, so without moving to named parameters (which I would still recommend), you could do something like:
```go
func postHandler(w http.ResponseWriter, r *http.Request) {
    if !len(r.URL.RawQuery) > 0 || !regexp.MustCompile(`^[a-zA-Z0-9]*$`).MatchString(r.URL.RawQuery) {
        fmt.Fprintf(w, "POST a key as an alphanumeric query with a JSON body\n")
        return
    }
    // Rest of the code ...
}
```

This way it's more obvious what preconditions need to be met, and the handling for them is up front.